system-contract-dapp-playground/src/api/hedera/tokenQuery-interactions/index.ts

```diff
@@ -0,0 +1,406 @@
+        );
+        break;
+
+      case 'KYC_STATUS':
+        transactionResult = await baseContract.getTokenDefaultKycStatusPublic(hederaTokenAddress);
+        break;
+
+      case 'CUSTOM_FEES':
+        transactionResult = await baseContract.getTokenCustomFeesPublic(hederaTokenAddress);
+        break;
+
+      case 'TOKEN_EXPIRY':
+        transactionResult = await baseContract.getTokenExpiryInfoPublic(hederaTokenAddress);
+        break;
+      case 'TOKEN_TYPE':
+        transactionResult = await baseContract.getTokenTypePublic(hederaTokenAddress);
+        break;
+      case 'TOKEN_KEYS':
+        if (!keyType) {
+          console.error('Key Type is needed for querying NON_FUNGIBLE_INFO');
+          return { err: 'Key Type is needed for querying NON_FUNGIBLE_INFO' };
+        } else {
+          transactionResult = await baseContract.getTokenKeyPublic(
+            hederaTokenAddress,
+            KEY_TYPE_MAP[keyType]
+          );
+        }
+        break;
+    }
+
+    return await handleContractResponseWithDynamicEventNames(transactionResult, eventMaps, API);
+  } catch (err: any) {
+    console.error(err);
+    return { err, transactionHash: err.receipt && err.receipt.hash };
+  }
+};
+
+/**
+ * @dev queries token's permission information
+ *
+ * @dev integrates TokenQueryContract.allowancePublic()
+ *
+ * @dev integrates TokenQueryContract.getApprovedPublic()
+ *
+ * @dev integrates TokenQueryContract.isApprovedForAllPublic()
+ *
+ * @param baseContract: ethers.Contract
+ *
+ * @param API: "ALLOWANCE" | "GET_APPROVED" | "IS_APPROVAL"
+ *
+ * @param hederaTokenAddress: string
+ *
+ * @param ownerAddress?: string
+ *
+ * @param spenderAddress?: string
+ *
+ * @param serialNumber?: number
+ *
+ * @return Promise<TokenQuerySmartContractResult>
+ */
+export const queryTokenPermissionInformation = async (
+  baseContract: Contract,
+  API: 'ALLOWANCE' | 'GET_APPROVED' | 'IS_APPROVAL',
+  hederaTokenAddress: string,
+  ownerAddress?: string,
+  spenderAddress?: string,
+  serialNumber?: number
+): Promise<TokenQuerySmartContractResult> => {
+  // sanitize param
+  if (!isAddress(hederaTokenAddress)) {
+    console.error('Invalid token address');
+    return { err: 'Invalid token address' };
+  } else if (ownerAddress && !isAddress(ownerAddress)) {
+    console.error('Invalid owner address');
+    return { err: 'Invalid owner address' };
+  } else if (spenderAddress && !isAddress(spenderAddress)) {
+    console.error('Invalid spender address');
+    return { err: 'Invalid spender address' };
+  } else if (serialNumber && serialNumber < 0) {
+    console.error('Invalid serial number');
+    return { err: 'Invalid serial number' };
+  }
+
+  // prepare events map
+  const eventMaps = {
+    IS_APPROVAL: 'Approved',
+    ALLOWANCE: 'AllowanceValue',
+    GET_APPROVED: 'ApprovedAddress',
+  };
+
+  let transactionResult, errMsg;
+  // invoking contract methods
+  try {
+    switch (API) {
+      case 'ALLOWANCE':
```

Comment:
Q: Are the error messages in here correct? They all talk about expiry but i believe they are related to approval or allowance

Reply:
Ah nice catch yeah sorry I must have forgotten to update them